storage/testonly/admin_storage_tester.go

```diff
@@ -579,3 +667,130 @@ func (tester *AdminStorageTester) TestAdminTXClose(t *testing.T) {
+	}
+	if !proto.Equal(got, want) {
+		return fmt.Errorf("post-getTree() diff (-got +want):\n%v", pretty.Compare(got, want))
+	}
+	return nil
+}
+
+type spec struct {
+	Tree            *trillian.Tree
+	Frozen, Deleted bool
+}
+
+// makeTreeOrFail delegates to makeTree. If makeTree returns a non-nil error, failFn is called.
+func makeTreeOrFail(ctx context.Context, failFn func(string, ...interface{}), s storage.AdminStorage, spec spec) *trillian.Tree {
+	tree, err := makeTree(ctx, s, spec)
+	if err != nil {
+		failFn("makeTree() returned err = %v", err)
+		return nil
+	}
+	return tree
+}
+
+// makeTree creates a tree and updates it to Frozen and/or Deleted, according to "spec".
+func makeTree(ctx context.Context, s storage.AdminStorage, spec spec) (*trillian.Tree, error) {
+	tree := proto.Clone(spec.Tree).(*trillian.Tree)
+
+	var err error
+	tree, err = createTree(ctx, s, tree)
+	if err != nil {
+		return nil, err
+	}
+
+	if spec.Frozen {
+		tree, _, err = updateTree(ctx, s, tree.TreeId, func(t *trillian.Tree) {
+			t.TreeState = trillian.TreeState_FROZEN
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if spec.Deleted {
+		tree, err = softDeleteTree(ctx, s, tree.TreeId)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Sanity checks
+	if spec.Frozen && tree.TreeState != trillian.TreeState_FROZEN {
+		return nil, fmt.Errorf("makeTree(): TreeState = %s, want = %s", tree.TreeState, trillian.TreeState_FROZEN)
+	}
+	if tree.Deleted != spec.Deleted {
+		return nil, fmt.Errorf("makeTree(): Deleted = %v, want = %v", tree.Deleted, spec.Deleted)
+	}
+
+	return tree, nil
+}
+
+func createTree(ctx context.Context, s storage.AdminStorage, tree *trillian.Tree) (*trillian.Tree, error) {
+	tx, err := s.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Close()
+	newTree, err := tx.CreateTree(ctx, tree)
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+	return newTree, nil
+}
+
+// updateTree updates the specified tree.
+// The bool return signifies whether the error was returned by the UpdateTree() call.
+func updateTree(ctx context.Context, s storage.AdminStorage, treeID int64, updateFunc func(*trillian.Tree)) (*trillian.Tree, bool, error) {
+	tx, err := s.Begin(ctx)
+	if err != nil {
+		return nil, false, err
+	}
+	defer tx.Close()
+	newTree, err := tx.UpdateTree(ctx, treeID, updateFunc)
+	if err != nil {
+		return nil, true, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, false, err
+	}
+	return newTree, false, nil
+}
+
+func getTree(ctx context.Context, s storage.AdminStorage, treeID int64) (*trillian.Tree, error) {
+	tx, err := s.Snapshot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Close()
+	tree, err := tx.GetTree(ctx, treeID)
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+	return tree, nil
+}
+
+func softDeleteTree(ctx context.Context, s storage.AdminStorage, treeID int64) (*trillian.Tree, error) {
```

Comment:
might be more consistent to move this function up to just after the 2 soft delete test cases

Reply:
There is a logic here: makeTree cares about softDelete, so I pushed it to the end of the file. A few other tests, like List and HardDelete care about soft-deleted trees too. OTOH, hardDelete and undelete are only called by their respective tests.

Reply:
Fair enough